File: queries/h11.py
import ibis


def tpc_h11(con, NATION='GERMANY', FRACTION=.0001):
    partsupp = con.table('partsupp')
    supplier = con.table('supplier')
    nation = con.table('nation')

    q = partsupp
    q = q.join(supplier, partsupp.PS_SUPPKEY == supplier.S_SUPPKEY)
    q = q.join(nation, nation.N_NATIONKEY == supplier.S_NATIONKEY)
    q = q.materialize()

    q = q.filter([q.N_NAME == NATION])

    innerq = partsupp
    innerq = innerq.join(supplier, partsupp.PS_SUPPKEY == supplier.S_SUPPKEY)
    innerq = innerq.join(nation, nation.N_NATIONKEY == supplier.S_NATIONKEY)
    innerq = innerq.materialize()
    innerq = innerq.filter([innerq.N_NAME == NATION])
    innerq = innerq.aggregate(total=(innerq.PS_SUPPLYCOST * innerq.PS_AVAILQTY).sum())

    gq = q.group_by([q.PS_PARTKEY])
    q = gq.aggregate(value=(q.PS_SUPPLYCOST * q.PS_AVAILQTY).sum())
    q = q.filter([q.value > innerq.total * FRACTION])
    q = q.sort_by(ibis.desc(q.value))
    return q

File: queries/h11.R
tpc_h11 <- function(input_func, collect_func = dplyr::collect) {
  nation <- input_func("nation") %>%
    filter(n_name == "GERMANY")

  joined_filtered <- input_func("partsupp") %>%
    inner_join(input_func("supplier"), by = c("ps_suppkey" = "s_suppkey")) %>%
    inner_join(nation, by = c("s_nationkey" = "n_nationkey"))

  global_agr <- joined_filtered %>%
    summarise(
      global_value = sum(ps_supplycost * ps_availqty) * 0.0001000000
    ) %>%
    mutate(global_agr_key = 1L)

  partkey_agr <- joined_filtered %>%
    group_by(ps_partkey) %>%
    summarise(value = sum(ps_supplycost * ps_availqty))

  partkey_agr %>%
    mutate(global_agr_key = 1L) %>%
    inner_join(global_agr, by = "global_agr_key") %>%
    filter(value > global_value) %>%
    arrange(desc(value)) %>%
    select(ps_partkey, value) %>%
    collect_func()
}
